Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs

```diff
@@ -0,0 +1,161 @@
+        /// <param name="valueFactory">A factory which returns a data</param>
+        /// <param name="shouldCache"></param>
+        /// <param name="dependenciesFactory"></param>
+        /// <returns>The data of generic type</returns>
+        public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> valueFactory, Func<T, bool> shouldCache = null, Func<T, IEnumerable<string>> dependenciesFactory = null)
+        {
+            if (await TryGetAsync(key, out T entry))
+            {
+                return entry;
+            }
+
+            var entryLock = _createLocks.GetOrAdd(key, _ => new SemaphoreSlim(1, 1));
+            try
+            {
+                await entryLock.WaitAsync();
+
+                if (await TryGetAsync(key, out entry))
+                {
+                    return entry;
+                }
+
+                var value = await valueFactory();
+
+                // Decide if the value should be cached based on the response
+                if (shouldCache != null && !shouldCache(value))
+                {
+                    return value;
+                }
+
+                // Set different timeout for stale content
+                var valueCacheOptions = new MemoryCacheEntryOptions();
+                if (value is AbstractResponse ar && ar.HasStaleContent)
+                {
+                    valueCacheOptions.SetAbsoluteExpiration(_cacheOptions.StaleContentExpiration);
+                }
+                else
+                {
+                    valueCacheOptions.SetSlidingExpiration(_cacheOptions.DefaultExpiration);
+                }
+
+                var dependencies = dependenciesFactory?.Invoke(value) ?? new List<string>();
+                var dependencyCacheOptions = new MemoryCacheEntryOptions { Priority = CacheItemPriority.NeverRemove };
+                foreach (var dependency in dependencies)
+                {
+                    var dependencyKey = dependency;
+                    var dependencyLock = _dependencyLocks.GetOrAdd(dependencyKey, _ => new object());
+
+                    if (!_memoryCache.TryGetValue(dependencyKey, out CancellationTokenSource tokenSource) || tokenSource.IsCancellationRequested)
+                    {
+                        lock (dependencyLock)
+                        {
+                            if (!_memoryCache.TryGetValue(dependencyKey, out tokenSource) || tokenSource.IsCancellationRequested)
+                            {
+                                tokenSource = _memoryCache.Set(dependencyKey, new CancellationTokenSource(), dependencyCacheOptions);
+                            }
+                        }
+                    }
+
+                    if (tokenSource != null)
+                    {
+                        valueCacheOptions.AddExpirationToken(new CancellationChangeToken(tokenSource.Token));
+                    }
+                }
+
+                return _memoryCache.Set(key, value, valueCacheOptions);
```

Comment:
@tomasjurasek 

Just out of curiosity - wouldn't it be better to use `_memoryCache.GetOrCreateAsync()`? 

- https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.cacheextensions.getorcreateasync?view=dotnet-plat-ext-3.1
- https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.1
- https://www.hanselman.com/blog/EyesWideOpenCorrectCachingIsAlwaysHard.aspx